### PR TITLE
Generate build info from environment variable

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

This PR is an alternative to PR #3362 in the way that procedural macro is not needed. It acquires build info and put it in environment variables, which is later retrieved in tikv binaries at build time. In this way, build.rs can be eliminated, while we also don't need to create a proc macro.

If users directly use cargo command to build tikv instead of Makefile, these fields will become "Unknown (...)".

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

Manually.

The output of `--version` from `make dev`:

```
$ target/debug/tikv-server --version
TiKV 
Release Version:   2.1.0-beta
Git Commit Hash:   7512f3cdb38670f44108396ff1e680bfb462414d
Git Commit Branch: wenxuan/speed_up_bench_modify
UTC Build Time:    2018-07-30 12:18:23
Rust Version:      rustc 1.29.0-nightly (4f3c7a472 2018-07-17)
```

The output of `--version` from `cargo build`:

```
$ target/debug/tikv-server --version
TiKV 
Release Version:   2.1.0-beta
Git Commit Hash:   Unknown (env var does not exist when building)
Git Commit Branch: Unknown (env var does not exist when building)
UTC Build Time:    Unknown (env var does not exist when building)
Rust Version:      Unknown (env var does not exist when building)
```

The output of `--version` from `make dev` when git folder does not exist:

```
$ target/debug/tikv-server --version
TiKV 
Release Version:   2.1.0-beta
Git Commit Hash:   Unknown (no git or not git repo)
Git Commit Branch: Unknown (no git or not git repo)
UTC Build Time:    2018-07-30 12:24:33
Rust Version:      rustc 1.29.0-nightly (4f3c7a472 2018-07-17)
```

The output of `--version` from `make dev` when `git` command is not available:

```
$ target/debug/tikv-server --version
TiKV 
Release Version:   2.1.0-beta
Git Commit Hash:   Unknown (no git or not git repo)
Git Commit Branch: Unknown (no git or not git repo)
UTC Build Time:    2018-07-30 12:29:45
Rust Version:      rustc 1.29.0-nightly (4f3c7a472 2018-07-17)
```

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Refer to a related PR or issue link (optional)

#3362

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -27,6 +27,13 @@ BIN_PATH = $(CURDIR)/bin
 GOROOT ?= $(DEPS_PATH)/go
 CARGO_TARGET_DIR ?= $(CURDIR)/target
 
+BUILD_INFO_GIT_FALLBACK := "Unknown (no git or not git repo)"
+BUILD_INFO_RUSTC_FALLBACK := "Unknown"
+export TIKV_BUILD_TIME := $(shell date -u '+%Y-%m-%d %I:%M:%S')
+export TIKV_BUILD_GIT_HASH := $(shell git rev-parse HEAD 2> /dev/null || echo ${BUILD_INFO_GIT_FALLBACK})
```
Comment (BusyJay) [inline Makefile]: Will tikv-server include the latest hash if I commit the local changes?

File: fuzz/fuzz_targets/coprocessor/codec/mysql/decimal.rs [hunk 2]
```diff
@@ -11,7 +11,7 @@ where
     I: Iterator<Item = u8>,
 {
     let mut bytes = [0u8; 8];
-    for byte in bytes.iter_mut() {
+    for byte in &mut bytes {
```
Comment (siddontang) [inline fuzz/fuzz_targets/coprocessor/codec/mysql/decimal.rs]: em, why include this in this PR?

**Comment by siddontang** [comment]
do you copy the binary to another machine and check the output?

File: fuzz/fuzz_targets/coprocessor/codec/mysql/decimal.rs (hunk 2 above)
Comment (breeswish) [reply fuzz/fuzz_targets/coprocessor/codec/mysql/decimal.rs]: this PR also makes clippy covers all workspaces (introduced in previous commits), and this is an issue that clippy reports.

File: Makefile (hunk 1 above)
Comment (breeswish) [reply Makefile]: If there is no source change after you committed local change, then the hash is not changed. Notice that this behavior is the same for current master:

```
# breezewish @ ~/Work/PingCAP/tikv on git:master x [13:45:16] C:130
$ target/debug/tikv-server --version
TiKV 
Release Version:   2.1.0-beta
Git Commit Hash:   8f1124735e591d255bde6a24bb4907aab1aa9c41
Git Commit Branch: master
UTC Build Time:    2018-07-31 05:43:34
Rust Version:      1.29.0-nightly (4f3c7a472 2018-07-17)

# breezewish @ ~/Work/PingCAP/tikv on git:master x [13:45:24] 
$ git commit --allow-empty -m "test"
[master c2c27d584] test

# breezewish @ ~/Work/PingCAP/tikv on git:master x [13:45:30] 
$ make dev
info: component 'rustfmt-preview' for target 'x86_64-apple-darwin' is up to date
info: component 'clippy-preview' for target 'x86_64-apple-darwin' is up to date
    Finished dev [unoptimized + debuginfo] target(s) in 0.89s                                                                                                                                                    
# When SIP is enabled, DYLD_LIBRARY_PATH will not work in subshell, so we have to set it
# again here. LOCAL_DIR is defined in .travis.yml.
export DYLD_LIBRARY_PATH=":/lib" && \
	export LOG_LEVEL=DEBUG && \
	export RUST_BACKTRACE=1 && \
	cargo test --features "default sse"  -- --nocapture && \
	cargo test --features "default sse" --bench benches  -- --nocapture  && \
	if [[ "`uname`" == "Linux" ]]; then \
		export MALLOC_CONF=prof:true,prof_active:false && \
		cargo test --features "default sse"  --bin tikv-server -- --nocapture --ignored; \
	fi
    Finished dev [unoptimized + debuginfo] target(s) in 2.49s

# breezewish @ ~/Work/PingCAP/tikv on git:master x [13:45:38] C:130
$ target/debug/tikv-server --version
TiKV 
Release Version:   2.1.0-beta
Git Commit Hash:   8f1124735e591d255bde6a24bb4907aab1aa9c41
Git Commit Branch: master
UTC Build Time:    2018-07-31 05:43:34
Rust Version:      1.29.0-nightly (4f3c7a472 2018-07-17)
```

**Comment by breeswish** [comment]
@busyjay @oveervenus @siddontang Please prioritize the review of this PR, since we need it to compare implementation performance during the Mvcc refinement. Thanks!

**Pull request merged**